pkg/util/nutanix.go

```
@@ -0,0 +1,320 @@
+type NutanixMachineProviderConfigBuilder struct {
+	withNutanixFailureDomains bool
+	currentIndex              int32
+	providerConfig            *machinev1.NutanixMachineProviderConfig
+}
+
+// Build returns the generated NutanixMachineProviderConfig.
+func (n *NutanixMachineProviderConfigBuilder) Build() *machinev1.NutanixMachineProviderConfig {
+	if n.providerConfig != nil {
+		return n.providerConfig
+	}
+
+	n.providerConfig = &machinev1.NutanixMachineProviderConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: machinev1.GroupVersion.String(),
+			Kind:       "NutanixMachineProviderConfig",
+		},
+		UserDataSecret:    &corev1.LocalObjectReference{Name: "nutanix-user-data"},
+		CredentialsSecret: &corev1.LocalObjectReference{Name: "nutanix-credentials"},
+		Image: machinev1.NutanixResourceIdentifier{
+			Type: machinev1.NutanixIdentifierName,
+			Name: ptr.To[string]("rhcos"),
+		},
+		Subnets:        []machinev1.NutanixResourceIdentifier{{Type: machinev1.NutanixIdentifierName, Name: ptr.To[string]("default-net")}},
+		VCPUsPerSocket: int32(1),
+		VCPUSockets:    int32(4),
+		MemorySize:     resource.MustParse(fmt.Sprintf("%dMi", 8096)),
+		Cluster: machinev1.NutanixResourceIdentifier{
+			Type: machinev1.NutanixIdentifierUUID,
+			UUID: ptr.To[string]("7244448a-7fde-400d-bf2e-bd8521459859"),
+		},
+		SystemDiskSize: resource.MustParse(fmt.Sprintf("%dGi", 120)),
+	}
+
+	if n.withNutanixFailureDomains {
+		fds := GetNutanixFailureDomains4Test()
+		UpdateNutanixMachineProviderConfigFromFailureDomain(n.providerConfig, &fds[n.currentIndex])
+	}
+
+	return n.providerConfig
+}
+
+// BuildRawExtension builds a new Nutanix machine config based on the configuration provided.
+func (n *NutanixMachineProviderConfigBuilder) BuildRawExtension() *runtime.RawExtension {
+	providerConfig := n.Build()
+
+	raw, err := json.Marshal(providerConfig)
+	if err != nil {
+		// As we are building the input to json.Marshal, this should never happen.
+		panic(err)
```

if this is used in non-test code, is there some way we could recover from the error?

i'm just concerned about having  a controller panic in the middle of execution instead of logging errors and continuing.

---

Yes, the struct NutanixMachineProviderConfigBuilder and NutanixFailureDomainsBuilder are used only for unit testing, as explaining in the comments.